src/test/app/AMM_test.cpp

```diff
@@ -0,0 +1,8180 @@
+
+        STAmount sa;
+        STPathSet paths;
+        if (result.isMember(jss::alternatives))
+        {
+            auto const& alts = result[jss::alternatives];
+            if (alts.size() > 0)
+            {
+                auto const& path = alts[0u];
+
+                if (path.isMember(jss::source_amount))
+                    sa = amountFromJson(sfGeneric, path[jss::source_amount]);
+
+                if (path.isMember(jss::destination_amount))
+                    da = amountFromJson(
+                        sfGeneric, path[jss::destination_amount]);
+
+                if (path.isMember(jss::paths_computed))
+                {
+                    Json::Value p;
+                    p["Paths"] = path[jss::paths_computed];
+                    STParsedJSONObject po("generic", p);
+                    paths = po.object->getFieldPathSet(sfPaths);
+                }
+            }
+        }
+
+        return std::make_tuple(std::move(paths), std::move(sa), std::move(da));
+    }
+    /************************************************/
+};
+
+struct AMM_test : public Test
+{
+public:
+    AMM_test() : Test()
+    {
+    }
+
+private:
+    void
+    testInstanceCreate()
+    {
+        testcase("Instance Create");
+
+        using namespace jtx;
+
+        // XRP to IOU
+        testAMM([&](AMM& ammAlice, Env&) {
+            BEAST_EXPECT(ammAlice.expectBalances(
+                XRP(10000), USD(10000), IOUAmount{10000000, 0}));
+        });
+
+        // IOU to IOU
+        testAMM(
+            [&](AMM& ammAlice, Env&) {
+                BEAST_EXPECT(ammAlice.expectBalances(
+                    USD(20000), BTC(0.5), IOUAmount{100, 0}));
+            },
+            {{USD(20000), BTC(0.5)}});
+
+        // IOU to IOU + transfer fee
+        {
+            Env env{*this};
+            fund(env, gw, {alice}, {USD(25000), BTC(0.625)}, Fund::All);
+            env(rate(gw, 1.25));
+            env.close();
+            AMM ammAlice(env, alice, USD(20000), BTC(0.5));
+            BEAST_EXPECT(ammAlice.expectBalances(
+                USD(20000), BTC(0.5), IOUAmount{100, 0}));
+            // 25,000 - 20,000(AMM) - 0.25*20,000=5,000(fee) = 0
+            BEAST_EXPECT(expectLine(env, alice, USD(0)));
+            // 0.625 - 0.5(AMM) - 0.25*0.5=0.125(fee) = 0
+            BEAST_EXPECT(expectLine(env, alice, BTC(0)));
+        }
+
+        // Require authorization is set, account is authorized
+        {
+            Env env{*this};
+            env.fund(XRP(30000), gw, alice);
+            env.close();
+            env(fset(gw, asfRequireAuth));
+            env.close();
+            env.trust(USD(30000), alice);
+            env.close();
+            env(trust(gw, alice["USD"](30000)), txflags(tfSetfAuth));
+            env.close();
+            env(pay(gw, alice, USD(10000)));
+            env.close();
+            AMM ammAlice(env, alice, XRP(10000), USD(10000));
```

Comment:
I don't see any assert or BEAST_EXPECT. Can you explain the intention of this code block?